libs/server/Objects/Hash/HashObjectImpl.cs

```diff
@@ -115,21 +115,39 @@ private void HashGetAll(ref ObjectInput input, ref SpanByteAndMemory output)
             {
                 if (respProtocolVersion < 3)
                 {
-                    while (!RespWriteUtils.WriteArrayLength(hash.Count * 2, ref curr, end))
+                    while (!RespWriteUtils.WriteArrayLength(Count() * 2, ref curr, end))
                         ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
                 }
                 else
                 {
-                    while (!RespWriteUtils.WriteMapLength(hash.Count, ref curr, end))
+                    while (!RespWriteUtils.WriteMapLength(Count(), ref curr, end))
                         ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
                 }
 
-                foreach (var item in hash)
+                if (HasExpirableItems())
                 {
-                    while (!RespWriteUtils.WriteBulkString(item.Key, ref curr, end))
-                        ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
-                    while (!RespWriteUtils.WriteBulkString(item.Value, ref curr, end))
-                        ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                    foreach (var item in hash)
+                    {
+                        if (IsExpired(item.Key))
+                        {
+                            continue;
+                        }
+
+                        while (!RespWriteUtils.WriteBulkString(item.Key, ref curr, end))
+                            ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                        while (!RespWriteUtils.WriteBulkString(item.Value, ref curr, end))
+                            ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                    }
+                }
+                else
+                {
+                    foreach (var item in hash)
+                    {
+                        while (!RespWriteUtils.WriteBulkString(item.Key, ref curr, end))
+                            ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                        while (!RespWriteUtils.WriteBulkString(item.Value, ref curr, end))
+                            ObjectUtils.ReallocateOutput(ref output, ref isMemory, ref ptr, ref ptrHandle, ref curr, ref end);
+                    }
```

Comment:
I had to do this after I found performance issues without this implementation.

> I have to remove the AsEnumerable method to remove the addition allocation
I have added the performance benchmark with and without this duplicate code in the above comments

Maybe I can try to use local functions to reduce the duplicate code but I have not seen local functions used anywhere else in the project.

Reply:
I have changed to use local functions, I didn't find any performance issue when using local functions

Reply:
Are you talking about perf regression for when there is no expiration? It seems there is some overlap between the condition in HashExpirableItems and IsExpired. Do you have to check expirationsTimes every time inside the for loop? Ideally, you check it once in the beginning and then just try to retrieve the item. If we don't have any expiration items, perf should be pretty similar to what we had before.
If not at least try to avoid checking expirationTimes is not null in both HasExpirableItems and IsExpired

Reply:
> Ideally, you check it once in the beginning and then just try to retrieve the item. If we don't have any expiration items, perf should be pretty similar to what we had before.

This is what I did but it has perf regression. Check the following commit where I reverted that code: https://github.com/microsoft/garnet/pull/864/commits/0b965463eeab0bffc4cd1b09ee845de7f3380148
![image](https://github.com/user-attachments/assets/6a43f0a2-5943-4c4e-b1a4-6caa0385483c)
![image](https://github.com/user-attachments/assets/be0c80d1-53c8-4912-bee5-c3387cd59307)

Reply:
This is completely different from what you have now. They do not compare in terms of performance. My comment was to avoid code duplication between the two for loops. The two for loops below differ only the if statement

![image](https://github.com/user-attachments/assets/9d1c40ba-eb94-4fbf-8131-43ace4e4431a)

Can you the following instead?
![image](https://github.com/user-attachments/assets/87360faa-614a-48d3-a708-0845ef65f080)

NOTE: HasExpirable and IsExpired share a common check (i.e. expirationTimes is not null). I think we don't have to execute that check, especially inside the for loop, since the data structure should not be changing while we read it. Unless I am missing something?!

![image](https://github.com/user-attachments/assets/2ba9a652-2a09-457e-9f5f-738285d3c2a6)

![image](https://github.com/user-attachments/assets/e828d99b-7f3d-4939-a15e-00d9485e0ca3)

Reply:
It was done outside only to avoid additional `if` checks inside the happy path (when there is no expiration for any fields), maybe dynamic PGO can avoid the overhead of having an addition if check inside the look. I will check the code as you suggested

Reply:
Done the changes, I didn't see any perf difference